keps/sig-multicluster/4322-cluster-inventory/README.md

```diff
@@ -0,0 +1,993 @@
+  resources per member cluster, and keeps their status up-to-date. Each
+  cluster manager SHOULD be identified with a unique name. Each cluster
+  inventory resource SHOULD be managed by only one cluster manager. Examples
+  of cluster manager are projects like OCM, Karmada, Clusternet or Azure
+  fleet manager.
+
+- **Cluster Inventory Consumer**: Users or third party tools that use
+  cluster inventory API for the purpose of workload distribution, operation
+  management etc. The consumer SHOULD not create this API resource.
+
+- **Member Cluster**: A kubernetes cluster that is managed by the cluster
+manager. A cluster manager SHOULD have sufficient permission to access
+the member cluster to fetch the information so it can update the status
+of the cluster inventory API resource.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1: Multicluster Workload Distribution
+
+In this scenario, the unified API acts as a foundation for multicluster
+scheduling across various clusters. This means that the API will provide
+a standardized way to distribute workloads across multiple clusters. For
+instance, workload distribution tools like GitOps or
+[Work API](https://github.com/kubernetes-sigs/work-api) can leverage this
+API to make informed decisions about which cluster is best suited to handle
+a particular workload.
+
+This could be based on factors such as the current load on the cluster,
+its capacity, and its proximity to the data source or end-users. For
+example, if a cluster is already running at high capacity, the API could
+direct the workload to a less busy cluster. This ensures optimal
+resource utilization and can significantly improve the performance and
+efficiency of applications running on the clusters.
+
+#### Story 2: Operations and Management
+
+For multi-cluster administrators, the unified API provides a
+comprehensive view of all the clusters under their management. This
+includes verifying their memberships, understanding their status,
+capacity, and healthiness.
+
+For instance, the API could provide real-time data about the CPU usage,
+memory consumption, network latency, and other key performance
+indicators of each cluster. If a cluster is nearing its capacity or
+experiencing issues, the API could alert the administrators, allowing
+them to take proactive measures to prevent downtime or performance
+degradation.
+
+Moreover, the API could also provide insights into the membership of
+each cluster, such as the number of nodes, their roles, and their
+current status. This can help administrators manage the clusters more
+effectively and ensure that they are always operating at optimal levels.
+
+#### Story 3: Transparent to Consumers
```

Comment:
This User Story is also a key point on why linking NFD is a good thing for the community.
Currently there are many end-users as well as projects that have NFD as their dependency to operate. E.g is the GPU-Operator from NVIDIA, the Confidentials-Container Operator, The Network-Operator, and many other infrastructure projects, rely on NFD for operating the cluster. So having the same information at the Cluster level, would be of relevance for projects like HyperShift(OpenShift RedHat) were they rely heavily on NFD API's

Reply:
agree on the high level. However, I think we need to adapt the NFD to CFD to work with clusterInventory as we cannot just lay out 5k node features in it. The cluster level scheduler also shouldn't need to know the feature of each node.

Reply:
@ryanzhang-oss that for sure doesn't work. But what do you think about adding a field for per-node features (like `nodeProperties` in addition to properties):

```yaml
nodeProperties:
   - name: node-1
     properties:
       - name: my-property
         value: foobar
```

This could then be used to expose per-node information if/when needed

Reply:
About: 

```
The cluster level scheduler also shouldn't need to know the feature of each node.
```

I am working with multiple stakeholders, not just at my ORG, but in academia and other cloud vendors, where we need this information for the scheduler. When it comes to specialised workloads like HPC (MPI) and AI/ML, the scheduling decision can not be done just with CPU/Memory of the entire cluster

Reply:
would you give more details on why per node info is needed? Why exposing them as a cluster-wide property is not sufficient for multiple cluster scheduling? Does it also need to select a certain node in the cluster in addition to selecting an appropriate cluster?

Reply:
For some MPI or AI/ML workloads, we need not only to know if the cluster is GPU capable, but if the nodes are connected with fast network, how many nodes share this feature, etc. 
We don't expect this information to be used to decide the final node from the central management cluster, but to make an informed decision on which cluster suits best our workload needs

Reply:
Aren't these can also be exposed as labels or cluster wide properties, e.g. using `ClusterProperty` API in https://github.com/kubernetes-sigs/about-api? It seems to me these can be defined as "the number of nodes in the cluster having fast network"?

Reply:
lgtm